// @looker/sdk-node/lib/esm/nodeSettings.js
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import * as fs from 'fs';
import * as ini from 'ini';
import { ApiConfigMap, ApiSettings, DefaultSettings, ValueSettings, sdkError, unquote } from '@looker/sdk-rtl';
export var getenv = function getenv(name) {
  var defaultValue = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : undefined;
  var val = process.env[name];
  return val === undefined ? defaultValue : val;
};
export var ApiConfig = contents => ini.parse(contents);
export var ApiConfigSection = (contents, section) => {
  var config = ApiConfig(contents);
  if (!section) {
    section = Object.keys(config)[0];
  }
  var settings = config[section];
  if (!settings) {
    throw new Error("No section named \"".concat(section, "\" was found"));
  }
  if (settings.api_version) {
    console.warn('api_version is no longer read from a configuration file by the SDK');
  }
  return settings;
};
export var readEnvConfig = envPrefix => {
  var values = {};
  var configMap = ApiConfigMap(envPrefix);
  Object.keys(configMap).forEach(key => {
    var envKey = configMap[key];
    if (process.env[envKey] !== undefined) {
      var val = unquote(process.env[envKey]);
      values[key] = val;
    }
  });
  return values;
};
export var readIniConfig = (fileName, envPrefix, section) => {
  var config = readEnvConfig(envPrefix);
  if (fileName && fs.existsSync(fileName)) {
    config = _objectSpread(_objectSpread({}, ApiConfigSection(fs.readFileSync(fileName, 'utf8'), section)), config);
  }
  Object.keys(config).forEach(key => {
    var val = config[key];
    if (typeof val === 'string') {
      config[key] = unquote(val);
    }
  });
  return config;
};
export class NodeSettings extends ApiSettings {
  constructor(envPrefix, contents, section) {
    var settings;
    if (contents) {
      if (typeof contents === 'string') {
        settings = ApiConfigSection(contents, section);
      } else {
        settings = contents;
      }
      settings = _objectSpread(_objectSpread({}, readEnvConfig(envPrefix)), settings);
    } else {
      settings = readEnvConfig(envPrefix);
    }
    super(_objectSpread(_objectSpread({}, DefaultSettings()), settings));
    _defineProperty(this, "envPrefix", void 0);
    _defineProperty(this, "section", void 0);
    this.section = section !== null && section !== void 0 ? section : '';
    this.envPrefix = envPrefix;
  }
  readConfig(_section) {
    return readEnvConfig(this.envPrefix);
  }
}
export class NodeSettingsIniFile extends NodeSettings {
  constructor(envPrefix) {
    var fileName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
    var section = arguments.length > 2 ? arguments[2] : undefined;
    if (fileName && !fs.existsSync(fileName)) {
      throw sdkError({
        message: "File ".concat(fileName, " was not found")
      });
    }
    fileName = fileName || './looker.ini';
    var config = readIniConfig(fileName, envPrefix, section);
    var settings = ValueSettings(config, envPrefix);
    super(envPrefix, settings, section);
    _defineProperty(this, "fileName", void 0);
    this.fileName = fileName;
  }
  readConfig(section) {
    section = section || this.section;
    return readIniConfig(this.fileName, this.envPrefix, section);
  }
}
